File: commons-parsers/src/main/java/xyz/wagyourtail/commons/parsers/impl/constant/parts/WholePart.java
package xyz.wagyourtail.commons.parsers.impl.constant.parts;

import lombok.val;
import xyz.wagyourtail.commons.core.reader.CharReader;
import xyz.wagyourtail.commons.core.reader.StringCharReader;
import xyz.wagyourtail.commons.parsers.Data;
import xyz.wagyourtail.commons.parsers.StringData;

import java.util.Collections;

public class WholePart extends StringData.OnlyRaw<Data.SingleContent<String>> {

    public WholePart(String rawContent) {
        super(rawContent, WholePart::getContentChecked);
    }

    public WholePart(CharReader<?> reader) {
        super(reader, WholePart::getContentChecked);
    }

    public static WholePart parse(String rawContent) {
        StringCharReader reader = new StringCharReader(rawContent);
        val number = new WholePart(reader);
        reader.expectEOS();
        return number;
    }

    public static WholePart unchecked(String rawContent) {
        return new WholePart(rawContent);
    }

    private static SingleContent<String> getContentChecked(CharReader<?> reader) {
        val first = reader.peek();
        if (first == '0') {
            throw reader.createException("Whole number cannot start with 0");
        }
        if (first < '0' || first > '9') {
            throw reader.createException("Not a whole number: " + (char) first);
        }

        return new SingleContent<>(reader.takeWhile(e -> e >= '0' && e <= '9'));
    }

}

File: commons-parsers-kt-kt/src/commonMain/kotlin/xyz/wagyourtail/commonskt/parsers/impl/constant/number/WholePart.kt
package xyz.wagyourtail.commonskt.parsers.impl.constant.number

import xyz.wagyourtail.commonskt.parsers.Data
import xyz.wagyourtail.commonskt.parsers.StringData
import xyz.wagyourtail.commonskt.reader.CharReader

class WholePart(rawContent: String) : StringData.OnlyRaw<Data.SingleContent<String>>(rawContent) {

    constructor(content: SingleContent<String>) : this(content.toString())

    companion object : StringDataBuilder<WholePart, SingleContent<String>> {

        override fun invoke(rawContent: CharReader<*>): WholePart {
            return WholePart(checkedBuildContent(rawContent))
        }

        override fun checkedBuildContent(reader: CharReader<*>): SingleContent<String> {
            val first = reader.peek()
            if (first == '0') {
                throw reader.createException("Whole number cannot start with 0");
            }
            if (first != null && first.isDigit()) {
                return SingleContent(reader.takeWhile { it.isDigit() })
            }

            throw reader.createException("Not a whole number: $first")
        }

    }

    override fun checkedBuildContent(reader: CharReader<*>): SingleContent<String> {
        return Companion.checkedBuildContent(reader)
    }

}